File: python/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        i = 0
        p = q = head
        while p:
            i += 1
            p = p.next
        i = i - n
        p, pre = head, None
        while i:
            pre = p
            p = p.next
            i -= 1
        if pre and p:
            t = p
            pre.next = p.next
            del t
            return head
            
        t = head
        head = head.next
        del t
        
        return head


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        mp = {}
        cnt = 0
        p = head
        while p:
            mp[cnt] = p
            cnt += 1
            p = p.next

        idx = cnt-n
        if idx==0:
            return head.next
        if idx==cnt-1:
            mp[idx-1].next = None
            return head
        mp[idx-1].next = mp[idx].next
        return head
                

File: go/src/leetcode/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.go
package main

import "fmt"

type ListNode struct {
	Val int
	Next *ListNode
}

func BuildList(a []int) *ListNode {
	dummy := &ListNode{}
	p := dummy
	for _, num := range a {
		p.Next = &ListNode{num, nil}
		p = p.Next
	}
	return dummy.Next
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	p := head
	for i := 0; i < n; i++ {
		p = p.Next
	}
	q := dummy
	for p != nil {
		q = q.Next
		p = p.Next
	}
	q.Next = q.Next.Next
	return dummy.Next
}

func print(head *ListNode) {
	for head != nil {
		fmt.Printf("%v -> ", head.Val)
		head = head.Next
	}
	fmt.Println()
}

func main() {
	fmt.Println(removeNthFromEnd(BuildList([]int{1,2,3,4,5}), 2))
	fmt.Println(removeNthFromEnd(BuildList([]int{1}), 1))
	print(removeNthFromEnd(BuildList([]int{1,2}), 2))
}
